src/wallet/wallet_tx_builder.cpp

```diff
@@ -177,25 +195,232 @@ SpendableInputs WalletTxBuilder::FindAllSpendableInputs(
+        CAmount dustThreshold,
+        CAmount targetAmount,
+        CAmount changeAmount)
+{
+    return InvalidFundsError(
+            spendable.Total(),
+            hasQuasiChange
+            // TODO: NEED TESTS TO EXERCISE THIS
+            ? InvalidFundsReason(QuasiChangeError(fee, dustThreshold))
+            : (changeAmount > 0 && changeAmount < dustThreshold
+               // TODO: we should provide the option for the caller to explicitly forego change
+               // (definitionally an amount below the dust amount) and send the extra to the
+               // recipient or the miner fee to avoid creating dust change, rather than prohibit
+               // them from sending entirely in this circumstance. (Daira disagrees, as this could
+               // leak information to the recipient or publicly in the fee.)
+               ? InvalidFundsReason(DustThresholdError(dustThreshold, changeAmount))
+               : InvalidFundsReason(InsufficientFundsError(targetAmount))));
+}
+
+static void
+AddChangePayment(Payments& resolvedPayments, const ChangeAddress& changeAddr, CAmount changeAmount)
+{
+    assert(changeAmount > 0);
+
+    examine(changeAddr, match {
+        // TODO: Once we can add Sprout change to `resolvedPayments`, we don’t need to pass
+        //       `changeAddr` around the rest of these functions.
+        [](const libzcash::SproutPaymentAddress&) {},
+        [](const libzcash::SproutViewingKey&) {},
+        [&](const auto& sendTo) {
+            resolvedPayments.AddPayment(
+                    ResolvedPayment(std::nullopt, sendTo, changeAmount, std::nullopt, true));
+        }
+    });
+}
+
+/// On the initial call, we haven’t yet selected inputs, so we assume the outputs dominate the
+/// actions.
+///
+/// 1. calc fee using only resolvedPayments to set a lower bound on the actual fee
+///    • this also needs to know which pool change is going to, so it can determine what the fee is
+///      with change _if_ there is change
+/// 2. iterate over LimitToAmount until the updated fee (now including spends) matches the expected
+///    fee
+tl::expected<
+    std::tuple<SpendableInputs, CAmount, std::optional<ChangeAddress>>,
+    InputSelectionError>
+WalletTxBuilder::IterateLimit(
+        CWallet& wallet,
+        const ZTXOSelector& selector,
+        const TransactionStrategy strategy,
+        CAmount sendAmount,
+        CAmount dustThreshold,
+        const SpendableInputs& spendable,
+        Payments& resolved,
+        bool afterNU5) const
+{
+    SpendableInputs spendableMut;
+
+    auto previousFee = MINIMUM_FEE;
+    auto updatedFee = CalcZIP317Fee(std::nullopt, resolved.GetResolvedPayments(), std::nullopt);
+    // This is used to increase the target amount just enough (generally by 0 or 1) to force
+    // selection of additional notes.
+    CAmount bumpTargetAmount{0};
+    std::optional<ChangeAddress> changeAddr;
+    CAmount changeAmount{0};
+
+    do {
+        spendableMut = spendable;
```

Comment:
Right – we need to do that to re-LimiToAmount. It’s my least favorite part of this change, through. Would you like a comment added about that?

Reply:
It'd be nice to have a comment, but only if you're going to make another CI run anyway; otherwise don't bother. I believe that @daira is currently re-reviewing this; if ze has changes to request then go ahead and change this too.